### ISSUE TITLE
MP3 support builds but doesn't work

### ISSUE THREAD

**Issue by hmijail** [libsndfile/libsndfile#852]
macOS Monterey 12.4 (on Apple Silicon)
libsndfile commit 53e7dee23435f16671e00a22ab1277624bae6f8e

Building with autotools fails because of a warning-turned-error: 
```
src/wavlike.c:838:11: error: variable 'bytes' set but not used [-Werror,-Wunused-but-set-variable]
        uint32_t bytes = 0 ;
```

Building with cmake works, and it reports that both lame and mpg123 were found and enabled. The resulting binaries are shown by `otool -L` to be linked against those libraries. And yet, sndfile-info and mpeg_test report that they have been compiled without mp3 support.

```
...
-- The following OPTIONAL packages have been found:

 * Mpg123 (required version >= 1.25.10)
...
-- The following RECOMMENDED packages have been found:
...
 * mp3lame, High quality MPEG Audio Layer III (MP3) encoder, <https://lame.sourceforge.io/>
   Enables MPEG layer III (MP3) writing support
```

```
./mpeg_test
    No MPEG tests because mpg123/lame support was not compiled in.

otool -L mpeg_test
mpeg_test:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
	/opt/homebrew/opt/libvorbis/lib/libvorbisenc.2.dylib (compatibility version 3.0.0, current version 3.12.0)
	/opt/homebrew/opt/libvorbis/lib/libvorbis.0.dylib (compatibility version 5.0.0, current version 5.9.0)
	/opt/homebrew/opt/flac/lib/libFLAC.8.dylib (compatibility version 12.0.0, current version 12.0.0)
	/opt/homebrew/opt/libogg/lib/libogg.0.dylib (compatibility version 9.0.0, current version 9.5.0)
	/opt/homebrew/opt/opus/lib/libopus.0.dylib (compatibility version 9.0.0, current version 9.0.0)
	/opt/homebrew/opt/mpg123/lib/libmpg123.0.dylib (compatibility version 48.0.0, current version 48.0.0)
	/opt/homebrew/opt/lame/lib/libmp3lame.0.dylib (compatibility version 1.0.0, current version 1.0.0)

```

**Comment by CarlGao4** [comment]
Please manually download [lame](https://sourceforge.net/projects/lame/files/lame/3.100/lame-3.100.tar.gz/download) and build install then rebuild libsndfile. It seems that lame is not included in brew.

**Comment by hmijail** [comment]
What do you mean that it is not included in brew? I have it installed and in use by ffmpeg.

```
$ brew ls lame
/opt/homebrew/Cellar/lame/3.100/bin/lame
/opt/homebrew/Cellar/lame/3.100/include/lame/lame.h
/opt/homebrew/Cellar/lame/3.100/lib/libmp3lame.0.dylib
/opt/homebrew/Cellar/lame/3.100/lib/ (2 other files)
/opt/homebrew/Cellar/lame/3.100/share/doc/ (14 files)
/opt/homebrew/Cellar/lame/3.100/share/man/man1/lame.1
```

**Comment by CarlGao4** [comment]
That's interesting, I'll try to find out this wierd behavior when I'm free.

**Comment by evpobr** [comment]
Hi @hmijail .  Strange. We are building test builds for mac OS 11.6.6, I looked again and there seem to be no problems:

```
checking for lame/lame.h... yes
checking for library containing lame_set_VBR_q... -lmp3lame
checking for libmpg123 >= 1.25.10 ... yes

    Experimental code : ................... no
    Using ALSA in example programs : ...... no (auto)
    External FLAC/Ogg/Vorbis/Opus : ....... yes
    External MPEG Lame/MPG123 : ........... yes
    Building Octave interface : ........... no
```

**Comment by evpobr** [comment]
@arthurt , any ideas?

**Comment by evpobr** [comment]
@hmijail , can you attach `CMakeCache.txt`, CMake and Autotools configure logs.

**Comment by hmijail** [comment]
[autoconf's config.log](https://github.com/libsndfile/libsndfile/files/9145270/config.log)
[CMakeCache.txt](https://github.com/libsndfile/libsndfile/files/9145293/CMakeCache.txt)

About CMake logs, do you mean the ones that appear in CMakeFiles? If so, here they are:
[CMakeError.log](https://github.com/libsndfile/libsndfile/files/9145299/CMakeError.log)
[CMakeOutput.log](https://github.com/libsndfile/libsndfile/files/9145300/CMakeOutput.log)

**Comment by hmijail** [comment]
In case it helps, I notice that the lame installation did not create a pkg-config file in /opt/homebrew/lib/pkgconfig/, while the mpg123 instalation did.

And indeed, config.log shows that pkg-config was used for libmpg123, but not for lame.

**Comment by evpobr** [comment]
> In case it helps, I notice that the lame installation did not create a pkg-config file in /opt/homebrew/lib/pkgconfig/, while the mpg123 instalation did.

The developers of LAME did not bother to support it.

**Comment by evpobr** [comment]
I didn't see anything useful in the CMake logs. Can you send the build log? Maybe there are useful warnings there.

**Comment by evpobr** [comment]
I suspect something is wrong LAME installation. Obsolete Homebrew recipe? Can you update all and reinstall LAME?

**Comment by hmijail** [comment]
> Can you send the build log?

Which file exactly? I don't see any more .log files anywhere.

> Obsolete Homebrew recipe? Can you update all and reinstall LAME?

I did `brew reinstall lame`. Files look the same. Tried `./configure` again, same result (mpg123 found, lame not).

**Comment by hmijail** [comment]
Looking at the configure log, looks like all libraries get an -I and -L parameter. Lame is the one which doesn't, and sure enough the header and lib aren't found. 
So I tried this and configure now finds everything:

```
CFLAGS='-I/opt/homebrew/include/ -L/opt/homebrew/lib/' ./configure
...

checking for lame/lame.h... yes
checking for library containing lame_set_VBR_q... -lmp3lame
checking for libmpg123 >= 1.25.10 ... yes
...
```

Additionally, `make` works (because of the missing -werror in configure) and now the sndfile-info binary works with a test mp3.

**Comment by hmijail** [comment]
I suspect that the problem is in that Homebrew on Intel uses /usr/local/{include,lib}, which conveniently are defaults for compilers, so anything installed is found automatically. 

But on Apple Silicon, Homebrew uses /opt/homebrew/{include, lib}, which aren't defaults and so nothing is found if not explicitly specified. Pkg-config helps for lots of cases... but not for lame.

**Comment by evpobr** [comment]
> Looking at the configure log, looks like all libraries get an -I and -L parameter. Lame is the one which doesn't, and sure enough the header and lib aren't found.

Correct library paths for other libraries are provided by PkgConfig:

```
Libs: -L${libdir} -lsndfile
flags: -I${includedir}
```

But LAME has no PkgConfig module.

Probably this could help:

https://stackoverflow.com/questions/7561509/how-to-add-include-and-lib-paths-to-configure-make-cycle

**Comment by evpobr** [comment]
Again, the question arises why our test macOS builds work without any problems.

**Comment by hmijail** [comment]
> Again, the question arises why our test macOS builds work without any problems.

Is your test macOS build made with brew or with lame installed manually?
Is it done on Intel or Apple Silicon?

**Comment by evpobr** [comment]
> Is your test macOS build made with lame installed manually or with brew?

brew

Hmm, i don't know: https://github.com/actions/virtual-environments/blob/macOS-11/20220627.1/images/macos/macos-11-Readme.md

**Comment by evpobr** [comment]
> Is your test macOS build made with lame installed manually or with brew?

By the way, I noticed now that we don't install LAME. It looks like it is already pre-installed in the image.

**Comment by hmijail** [comment]
Seems to be Intel, because there's references to x64 arch, but no arm.

**Comment by hmijail** [comment]
In any case, if your lame lib and headers are in the compiler default paths, that explains why it works.

**Comment by hmijail** [comment]
> By the way, I noticed now that we don't install LAME. It looks like it is already pre-installed in the image.

The image also mentions homebrew and sox, for whom brew installs lame as a dependency. 

Given that it's x64 arch, to me this sounds again like the case of lame installed by brew on /usr/local/lib and /usr/local/include, and therefore found by default by the compiler.

**Comment by evpobr** [comment]
So we see the classic problem of finding libraries without a package manager.

@hmijail , i'm not Autotools export, but this could work:

```
$ configure --extra-ldflags="-L/path/to/extra/lib" --extra-cflags="-I/path/to/extra/includes"
```

**Comment by hmijail** [comment]
`configure` didn't like `--extra-ldflags` nor `--extra-cflags`. But it worked like this:
```
./configure LDFLAGS="-L/opt/homebrew/lib" CFLAGS="-I/opt/homebrew/include"
```

**Comment by evpobr** [comment]
It's good that you found a solution. But this is a problem when package managers are not used. I don't think there is a way to help this from our side.

**Comment by hmijail** [comment]
I agree that the problem of configure not finding libs and headers in non-standard paths is wider and typical and you can't solve it; I might raise that with a package manager.

But please remember that this issue was initially about cmake: it does find lame, compiles correctly, the binaries are linked against it, and yet binaries report that "mpg123/lame support was not compiled in".

**Comment by NeGitR** [comment]
> I agree that the problem of configure not finding libs and headers in non-standard paths is wider and typical and you can't solve it; I might raise that with a package manager.
> 
> But please remember that this issue was initially about cmake: it does find lame, compiles correctly, the binaries are linked against it, and yet binaries report that "mpg123/lame support was not compiled in".

Have the same issue !

**Comment by evpobr** [comment]
Hi @NeGitR . Could you provide more information please.

**Comment by luzpaz** [comment]
Is this still relevant ?
